docs/proposals/hpa-custom-metrics.md

```diff
@@ -0,0 +1,162 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<!-- TAG RELEASE_LINK, added by the munger automatically -->
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.3/docs/proposals/custom-metrics.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+Specifying Metrics in the HPA
+=============================
+
+Introduction
+------------
+
+Currently, custom metrics specified via a list of metric names
+and target values.  Each metric name is prefixed with "custom/",
+and then is requested at the pod level for each pod of the target
+scalable.
+
+Several use cases are not supported by this design:
+
+### Refering to Non-Custom Metrics ###
+
+Several users have asked for the ability to scale on existing built-in metrics
+(e.g. memory usage).  While the current design does allow the user to specify
+a metric name, all metrics are prefixed with "custom/", meaning built-in
+metrics cannot be passed this way.  A new design should not have completely
+separate mechanisms for custom and non-custom metrics.  This also allows
+refering to raw CPU usage rate, instead of refering to CPU usage as a
+percentage of the CPU request (this has been a commonly asked-for feature
+by users and operators).
+
+### Refering to Metrics Not Associated With Pods ###
+
+Heapster current can track metrics associated with pods and namespaces.  With
+Heapster push metrics, it becomes possible to define custom metrics on the
+namespace level.  However, the custom metrics annotation in the HPA currently
+only allows referring to custom metrics defined on pods.
+
+Refering to custom metrics at the namespace level allows us to support certain
+use cases in which an application wishes to expose a certain special metric
+to an entire namespace (e.g. queue length) and scale based on that.
+Additionally, until Heapster has the capability to track metrics from service
+objects, RCs, etc, namespace-level metrics can be used with push metrics to
+"fake" this type of metrics.  Metrics associated with services could be
+extracted from a load balancer or reverse proxy, for instance.
```

Comment:
> I don't see any reason why this mapping should happen implicitly in HPA controller and not externally in a component which is writing this metric.

I specifically made it *not* implicit.  You explicitly write "scale when service/foo > bar", so the HPA controller does not need to know anything about the relationship between the service and the RC/Deployment/etc.

Furthermore, I *do not* think it is a good idea to force metric writers to need to know how to associate a service with an RC (I don't think we should force metrics writers to do more work than is necessary, and pull in watches on other objects just to push metrics), and I also do not think it's a good idea to store metrics that are actually associated with a service as associated with a RC just to autoscale based on them.  Don't force a metric writer to say "well, someone *might* want to scale based on this metric, so we either have to store it as associated with the some other RC/Deployment/etc, and/or double store it as associated with both the service and the RC/Deployment/etc".  That seems broken to me.